### ISSUE TITLE
Inconsistent naming of size/length methods

### ISSUE THREAD

**Issue by dgelessus** [kaitai-io/kaitai_struct#527]
For [strings](http://doc.kaitai.io/user_guide.html#_strings) and [byte arrays](http://doc.kaitai.io/user_guide.html#_byte_arrays) it's called `length`, but for [custom arrays](http://doc.kaitai.io/user_guide.html#_array_types) (created with `repeat`) it's called `size`. Is there a reason for the different names? To me it seems like `length` and `size` are do the same thing (just for different types), so they should be named the same as well.

**Comment by GreyCat** [comment]
It was more or less the conscious choice back when I was introducing them, to emphasize difference between true arrays and strings/byte arrays.

We can review it and change, probably after 1.x — see #55.

If you really want to make that change, I'd start with analyzing of how these methods are named in existing languages that people are used to.

**Comment by dgelessus** [comment]
> We can review it and change, probably after 1.x

Yeah, this is not a high-priority issue obviously, so I'm fine with waiting 😃

> If you really want to make that change, I'd start with analyzing of how these methods are named in existing languages that people are used to.

In the languages that I know a little:

* Clojure, Go, Groovy, Lua, Python, Rust, and Swift use one length function/method/property/operator for all collections and strings
  * Go has both `len(obj)` and `obj.Len()`, but that's probably for technical reasons, and the names are basically the same
* C++ uses `coll.size()` for STL collections and strings, but strings also support `str.length()`
* Java, Kotlin, Objective-C, PHP, and Prolog have a different length function/method for strings than for other collections
  * For Kotlin and Prolog this likely an intentional design choice, for the others it might be a historical artifact
* C only has `strlen` and provides almost no support for other dynamic collections

Interestingly, Groovy only has `.size()` and Swift only has `.count`, even though their "ancestors" (Java and Objective-C) have multiple length operations. Kotlin however kept Java's separate `length` for strings and `size` for collections (but also changed it from `length` to `size` for arrays).

By the way, I don't care much about the exact name, as long as it's the same for all types. In other languages, `length`/`len` and `size` are roughly equally common, so I'd say choose whatever you like better 😃

**Comment by GreyCat** [comment]
@dgelessus Great table! Could I actually ask you to compile a table on who-uses-what exactly, so we could learn and use the most common spelling/name?

I can add:

* Ruby — allows usage of `.length` and `.size` on both strings & true arrays
* C with Glib — both GString and GArray has `.len` as a member of the structure

**Comment by dgelessus** [comment]
* Bash: `${#str}`, `${#arr[@]}`
* C (standard): `strlen(str)` (and `sizeof(type_or_obj)`, but that's not really a length operator)
* Clojure: `(count obj)`
* Go: `len(obj)`, `obj.Len()`
* Groovy: `obj.size()`
* Java: `arr.length`, `charseq.length()`, `coll.size()`
* Kotlin: `coll.size`, `charseq.length`
* Lua: `#obj`
* Objective-C: `[coll count]`, `[str length]`
* PHP: `count(arr)`/`sizeof(arr)` (`sizeof` is equivalent to `count` and is not like in C), `strlen(str)`
* Prolog: `length(Array, Length)`, `atom_length(Atom, Length)`, `string_length(String, Length)` (actual strings are a SWI-Prolog extension)
* Python: `len(obj)`
* Rust: `obj.len()`
* Swift: `obj.count`

Together with your list above, that makes:

* `length`: C++, C#, Java, Kotlin, Objective-C, Prolog, Ruby (7)
  * Of these, C++, C#, Kotlin, and Objective-C use `length` only for strings (but not collections)
* `len`: Glib, Go, Python, Rust (4)
  * `strlen`: C, PHP (2)
* `size`: C++, Groovy, Java, Kotlin, Ruby (5)
  * `sizeof`: C, PHP (2)
  * Of these, Java and Kotlin use `size` only for collections (but not strings)
* `count`: C#, Clojure, Objective-C, PHP, Qt, Swift (6)
  * Of these, C#, Objective-C, and PHP use `count` only for collections (but not strings)
* An operator: Bash, Lua (2)

So it seems that `length` is the most common, but it is often used only for strings - Prolog and Ruby are the only languages that use `length` for non-strings. `count` and `size` are the next most common choices, but they are also sometimes only used for collections and not strings. `len` seems to be a common choice in languages with a single length operation for all types.

**Comment by dgelessus** [comment]
Also, a Kaitai-specific argument against using `size` for "number of elements in array/string/collection": `size` is also used in specs to mean "byte count". I don't know if you can even add an explicit `size: ...` to a `repeat` array, but if you can, then it means something different than the array's `.size`.

Perhaps a better example is this from the user guide:

```yaml
seq:
  - id: my_len
    type: u4
  - id: my_str
    type: str
    size: my_len * 2
    encoding: UTF-16LE
```

If we used `.size` for all types, including strings, then `my_str.size` would be `my_len`, even though in the spec it says `size: my_len * 2`.

**Comment by KOLANICH** [comment]
It has been proposed to use sizeof for sizes of multiple elements, s.a. arrays in #84
